### ISSUE TITLE
Difficult bug when colnames are arrays

### ISSUE THREAD

**Issue by LTLA** [Bioconductor/SummarizedExperiment#7]
This bug was hell to track down:

```r
a <- matrix(10, 10, 26)
dimnames(a)[[2]] <- array(LETTERS, 26) 
se <- SummarizedExperiment(list(counts=a))
## Error in FUN(X[[i]], ...) : 
##  assay colnames() must be NULL or equal colData rownames()
```
`colnames(a)` is an array but gets converted to a character vector in the `SummarizedExperiment()` constructor. This results in the error as the array is not identical to the vector, presumably. However, it was very difficult to identify, not least because an array with 1 dimension looks identical to a vector (only by checking `dim()` did I figure it out). This is especially dangerous if anyone is manually pulling out dimension names from a HDF5 file, as _rhdf5_ automatically seems to pull things out as arrays.

I'm testing this on BioC-release (1.8.1) as I'm not at my other computer right now, but I guess this would be the same for the current devel?

**Comment by LTLA** [comment]
Bumping this, because another of my colleagues ran into this a few days ago.

**Comment by LTLA** [comment]
Probably option 1 sounds safest. Names on the character vector were also causing me problems with `SummarizedExperiment()` at one point, though that was much easier to spot.

More generally, enforcing de-attributed column names in ordinary matrices would be sensible. Though this would require a  fix to _Matrix_ as well, to get consistent behaviour across common matrix types.

**Comment by hpages** [comment]
Yes, better let the user know that something is wrong with the input so s/he can fix it.

Fixed in SummarizedExperiment 1.11.2 (https://github.com/Bioconductor/SummarizedExperiment/commit/f5bfedf5a70fe14042beb13a7ad289f8a1053d19) and ported to the RELEASE_3_7 branch (version 1.10.1). Note that the fix is not exactly option 1 discussed above: I only touched the `SummarizedExperiment()` constructor. The `names()`, `rownames()`, `colnames()`, and `dimnames()` setters actually accept insane right values but drop their attributes. This is consistent with what these setters do on most objects in base R.
